plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java

```diff
@@ -27,65 +27,124 @@
-        String vaultToken = getVariable("CASC_VAULT_TOKEN", prop);
-        String vaultAppRole = getVariable("CASC_VAULT_APPROLE", prop);
-        String vaultAppRoleSecret = getVariable("CASC_VAULT_APPROLE_SECRET", prop);
-        String vaultNamespace = getVariable("CASC_VAULT_NAMESPACE", prop);
-        String vaultEngineVersion = getVariable("CASC_VAULT_ENGINE_VERSION", prop);
-
-        if(((vaultPw != null && vaultUsr != null) || 
-            vaultToken != null || 
-            (vaultAppRole != null && vaultAppRoleSecret != null)) && vaultPth != null && vaultUrl != null) {
-            LOGGER.log(Level.FINE, "Attempting to connect to Vault: {0}", vaultUrl);
+
+        // attempt AppRole login
+        if (vaultAppRole.isPresent() && vaultAppRoleSecret.isPresent() && !authToken.isPresent()) {
             try {
-                VaultConfig config = new VaultConfig().address(vaultUrl);
-                if (vaultNamespace != null) {
-                    // optionally set namespace
-                    config = config.nameSpace(vaultNamespace);
-                    LOGGER.log(Level.FINE, "Using namespace with Vault: {0}", vaultNamespace);
-                }
-                if (vaultEngineVersion != null) {
-                    // optionally set vault engine version
-                    config = config.engineVersion( Integer.parseInt(vaultEngineVersion) );
-                    LOGGER.log(Level.FINE, "Using engine version: {0}", vaultEngineVersion);
-                }
-                config = config.build();
-                Vault vault = new Vault(config);
-                //Obtain a login token
-                final String token;
-                if (vaultToken != null) {
-                    token = vaultToken;
-                    LOGGER.log(Level.FINE, "Using supplied token to access Vault");
-                } else if (vaultAppRole != null && vaultAppRoleSecret != null) {
-                    token = vault.auth().loginByAppRole(vaultAppRole, vaultAppRoleSecret).getAuthClientToken();
-                    LOGGER.log(Level.FINE, "Login to Vault using AppRole/SecretID successful");
-                } else {
-                    token = vault.auth().loginByUserPass(vaultUsr, vaultPw, vaultMount).getAuthClientToken();
-                    LOGGER.log(Level.FINE, "Login to Vault using U/P successful");
+                authToken = Optional.of(
+                        vault.auth().loginByAppRole(vaultAppRole.get(), vaultAppRoleSecret.get()).getAuthClientToken()
+                );
+                LOGGER.log(Level.FINE, "Login to Vault using AppRole/SecretID successful");
+            } catch (VaultException e) {
+                LOGGER.log(Level.WARNING, "Could not login with AppRole", e);
+            }
+        }
+
+        // attempt User/Pass login
+        if (vaultUser.isPresent() && vaultPw.isPresent() && !authToken.isPresent()) {
+            try {
+                authToken = Optional.of(
```

Comment:
it returns a String that could be null, though I don't know of any cases it would be null since it would have thrown an exception beforehand
```suggestion
                authToken = Optional.ofNullable(
```

Reply:
You are right. Better safe than sorry here. The extra caution doesn't hurt